v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/postgresql/PostgreSQLDialectAdapter.java

```diff
@@ -0,0 +1,495 @@
+  public String getBoundaryQuery(
+      String tableName, ImmutableList<String> partitionColumns, String colName) {
+    return addWhereClause(
+        String.format("SELECT MIN(%s), MAX(%s) FROM %s", colName, colName, tableName),
+        partitionColumns);
+  }
+
+  /**
+   * Check if a given {@link SQLException} is a timeout. The implementation needs to check for
+   * dialect specific {@link SQLException#getSQLState() SqlState} to check if the exception
+   * indicates a server side timeout. The client side timeout would be already checked for by
+   * handling {@link SQLTimeoutException}, so the implementation does not need to check for the
+   * same.
+   */
+  @Override
+  public boolean checkForTimeout(SQLException exception) {
+    return exception.getSQLState() != null
+        && TIMEOUT_SQL_STATES.contains(exception.getSQLState().toUpperCase());
+  }
+
+  /**
+   * Get Query that returns order of collation. The query must return all the characters in the
+   * character set with the columns listed in {@link CollationOrderRow.CollationsOrderQueryColumns}.
+   *
+   * @param dbCharset character set used by the database for which collation ordering has to be
+   *     found.
+   * @param dbCollation collation set used by the database for which collation ordering has to be
+   *     found.
+   */
+  @Override
+  public String getCollationsOrderQuery(String dbCharset, String dbCollation) {
+    // TODO(thiagotnunes)
+    return "";
+  }
+
+  private String addWhereClause(String query, ImmutableList<String> partitionColumns) {
+    StringBuilder queryBuilder = new StringBuilder();
+    queryBuilder.append(query);
+    if (!partitionColumns.isEmpty()) {
+      queryBuilder.append(" WHERE ");
+      queryBuilder.append(
+          partitionColumns.stream()
+              // Include the column / range to define the where clause.
+              // `(exclude col = FALSE) OR (col >= range.start() AND (col < range.end() OR
+              // (range.isLast() = TRUE AND col = range.end()))`
+              .map(
+                  partitionColumn ->
+                      String.format(
+                          "((? = FALSE) OR (%1$s >= ? AND (%1$s < ? OR (? = TRUE AND %1$s = ?))))",
+                          partitionColumn))
+              .collect(Collectors.joining(" AND ")));
+    }
+    return queryBuilder.toString();
+  }
+
+  /**
+   * Ref <a
+   * href="https://www.postgresql.org/docs/16/catalog-pg-type.html#CATALOG-TYPCATEGORY-TABLE"></a>.
+   */
+  private SourceColumnIndexInfo.IndexType indexTypeFrom(String typeCategory) {
+    switch (typeCategory) {
+      case "N":
+        return SourceColumnIndexInfo.IndexType.NUMERIC;
+      case "D":
+        return SourceColumnIndexInfo.IndexType.DATE_TIME;
+      case "S":
+        return SourceColumnIndexInfo.IndexType.STRING;
+      default:
+        return SourceColumnIndexInfo.IndexType.OTHER;
+    }
+  }
+
+  private boolean shouldPadSpace(@Nullable String pad) throws SQLException {
```

Comment:
In PG, the `pad_attribute` whic is passed to this function is always "NO PAD". [Ref](https://www.postgresql.org/docs/16/infoschema-collations.html)

But

Whether trailing spaces are semantically significant or not, depends on the type of the column
[Ref](https://www.postgresql.org/docs/current/datatype-character.html)

Could we also consider the column type also in this logic?

Reply:
Done